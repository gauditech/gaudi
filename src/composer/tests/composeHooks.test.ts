import { compileToOldSpec, compose } from "@src/index";

describe("compose hooks", () => {
  it("composes source hooks", () => {
    const bp = `
      runtime MyRuntime {
        source path "some/path/to/file"
      }

      model Org {
        field name {
          type string,
          validate {
            hook {
              default arg name
              runtime MyRuntime
              source someHook from "githubc.js"
            }
          }
        }
        hook description {
          runtime MyRuntime
          source someHook from "githubc.js"
        }
      }

      entrypoint Orgs {
        target Org as org
        create endpoint {
          action {
            create {
              set name hook {
                runtime MyRuntime
                source someHook from "hooks.js"
              }
            }
          }
        }
      }

    `;
    const result = compose(compileToOldSpec(bp));

    expect(result).toMatchSnapshot();
  });

  it("composes inline hooks", () => {
    const bp = `
      runtime MyRuntime {
        source path "some/path/to/file"
      }

      model Org {
        field name {
          type string,
          validate {
            hook {
              default arg name
              inline "'test name'"
            }
          }
        }
        hook description {
          runtime MyRuntime
          inline "'some description'"
        }
      }

      entrypoint Orgs {
        target Org as org
        create endpoint {
          action {
            create {
              set name hook {
                inline "'test name'"
              }
            }
          }
        }
      }

    `;
    const result = compose(compileToOldSpec(bp));

    expect(result).toMatchSnapshot();
  });

  it("defaults to the default execution runtime when hook runtime is empty", () => {
    const bp = `
      runtime MyRuntime {
        default
        source path "some/path/to/file"
      }

      runtime MyRuntime2 {
        source path "some/path/to/file"
      }

      model Org { field name { type string } }

      entrypoint Orgs {
        target Org as org
        create endpoint {
          action {
            create {
              set name hook {
                source randomSlug from "hooks.js"
              }
            }
          }
        }
      }

    `;

    const result = compose(compileToOldSpec(bp));

    expect(result.entrypoints[0].endpoints[0]).toMatchSnapshot();
  });

  it("fails on invalid runtime name", () => {
    const bp = `
      runtime MyRuntime {
        source path "some/path/to/file"
      }

      entrypoint Orgs {
        target Org as org
        create endpoint {
          action {
            create {
              set name hook {
                runtime InvalidMyRuntime
                source randomSlug from "hooks.js"
              }
            }
          }
        }
      }

    `;

    expect(() => compose(compileToOldSpec(bp))).toThrowErrorMatchingInlineSnapshot(
      `"Unknown refkey: Org"`
    );
  });
});
