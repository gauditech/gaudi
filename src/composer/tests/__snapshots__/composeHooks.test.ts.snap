// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compose hooks composes inline hooks 1`] = `
{
  "entrypoints": [
    {
      "endpoints": [
        {
          "actions": [
            {
              "alias": "org",
              "changeset": [
                {
                  "name": "name",
                  "setter": {
                    "args": [],
                    "hook": {
                      "code": {
                        "inline": "'test name'",
                        "kind": "inline",
                      },
                      "runtimeName": "MyRuntime",
                    },
                    "kind": "fieldset-hook",
                  },
                },
              ],
              "kind": "create-one",
              "model": "Org",
              "select": [],
              "targetPath": [
                "org",
              ],
            },
          ],
          "authSelect": [],
          "authorize": undefined,
          "fieldset": {
            "kind": "record",
            "nullable": false,
            "record": {},
          },
          "kind": "create",
          "parentContext": [],
          "response": [
            {
              "alias": "id",
              "kind": "field",
              "name": "id",
              "namePath": [
                "Org",
                "id",
              ],
              "refKey": "Org.id",
            },
            {
              "alias": "name",
              "kind": "field",
              "name": "name",
              "namePath": [
                "Org",
                "name",
              ],
              "refKey": "Org.name",
            },
          ],
          "target": {
            "alias": "org",
            "kind": "model",
            "name": "Org",
            "namePath": [
              "Org",
            ],
            "refKey": "Org",
            "retType": "Org",
            "select": [],
          },
        },
      ],
      "entrypoints": [],
      "name": "Orgs",
      "target": {
        "alias": "org",
        "identifyWith": {
          "name": "id",
          "paramName": "org_id",
          "refKey": "Org.id",
          "type": "integer",
        },
        "kind": "model",
        "name": "Org",
        "namePath": [
          "Org",
        ],
        "refKey": "Org",
        "retType": "Org",
      },
    },
  ],
  "models": [
    {
      "aggregates": [],
      "computeds": [],
      "dbname": "org",
      "fields": [
        {
          "dbname": "id",
          "dbtype": "serial",
          "kind": "field",
          "modelRefKey": "Org",
          "name": "id",
          "nullable": false,
          "primary": true,
          "refKey": "Org.id",
          "type": "integer",
          "unique": true,
          "validators": [],
        },
        {
          "dbname": "name",
          "dbtype": "text",
          "kind": "field",
          "modelRefKey": "Org",
          "name": "name",
          "nullable": false,
          "primary": false,
          "refKey": "Org.name",
          "type": "text",
          "unique": false,
          "validators": [
            {
              "arg": "name",
              "hook": {
                "code": {
                  "inline": "'test name'",
                  "kind": "inline",
                },
                "runtimeName": "MyRuntime",
              },
              "name": "hook",
            },
          ],
        },
      ],
      "hooks": [
        {
          "args": [],
          "hook": {
            "code": {
              "inline": "'some description'",
              "kind": "inline",
            },
            "runtimeName": "MyRuntime",
          },
          "kind": "model-hook",
          "name": "description",
          "refKey": "Org.description",
        },
      ],
      "kind": "model",
      "name": "Org",
      "queries": [],
      "refKey": "Org",
      "references": [],
      "relations": [],
    },
  ],
  "populators": [],
  "resolveOrder": [
    "Org",
    "Org.name",
    "Org.description",
  ],
  "runtimes": [
    {
      "default": true,
      "name": "MyRuntime",
      "sourcePath": "some/path/to/file",
      "type": "node",
    },
    {
      "default": false,
      "name": "$GAUDI_INTERNAL",
      "sourcePath": "./internalExecutionRuntime",
      "type": "node",
    },
  ],
}
`;

exports[`compose hooks composes source hooks 1`] = `
{
  "entrypoints": [
    {
      "endpoints": [
        {
          "actions": [
            {
              "alias": "org",
              "changeset": [
                {
                  "name": "name",
                  "setter": {
                    "args": [],
                    "hook": {
                      "code": {
                        "file": "hooks.js",
                        "kind": "source",
                        "target": "someHook",
                      },
                      "runtimeName": "MyRuntime",
                    },
                    "kind": "fieldset-hook",
                  },
                },
              ],
              "kind": "create-one",
              "model": "Org",
              "select": [],
              "targetPath": [
                "org",
              ],
            },
          ],
          "authSelect": [],
          "authorize": undefined,
          "fieldset": {
            "kind": "record",
            "nullable": false,
            "record": {},
          },
          "kind": "create",
          "parentContext": [],
          "response": [
            {
              "alias": "id",
              "kind": "field",
              "name": "id",
              "namePath": [
                "Org",
                "id",
              ],
              "refKey": "Org.id",
            },
            {
              "alias": "name",
              "kind": "field",
              "name": "name",
              "namePath": [
                "Org",
                "name",
              ],
              "refKey": "Org.name",
            },
          ],
          "target": {
            "alias": "org",
            "kind": "model",
            "name": "Org",
            "namePath": [
              "Org",
            ],
            "refKey": "Org",
            "retType": "Org",
            "select": [],
          },
        },
      ],
      "entrypoints": [],
      "name": "Orgs",
      "target": {
        "alias": "org",
        "identifyWith": {
          "name": "id",
          "paramName": "org_id",
          "refKey": "Org.id",
          "type": "integer",
        },
        "kind": "model",
        "name": "Org",
        "namePath": [
          "Org",
        ],
        "refKey": "Org",
        "retType": "Org",
      },
    },
  ],
  "models": [
    {
      "aggregates": [],
      "computeds": [],
      "dbname": "org",
      "fields": [
        {
          "dbname": "id",
          "dbtype": "serial",
          "kind": "field",
          "modelRefKey": "Org",
          "name": "id",
          "nullable": false,
          "primary": true,
          "refKey": "Org.id",
          "type": "integer",
          "unique": true,
          "validators": [],
        },
        {
          "dbname": "name",
          "dbtype": "text",
          "kind": "field",
          "modelRefKey": "Org",
          "name": "name",
          "nullable": false,
          "primary": false,
          "refKey": "Org.name",
          "type": "text",
          "unique": false,
          "validators": [
            {
              "arg": "name",
              "hook": {
                "code": {
                  "file": "githubc.js",
                  "kind": "source",
                  "target": "someHook",
                },
                "runtimeName": "MyRuntime",
              },
              "name": "hook",
            },
          ],
        },
      ],
      "hooks": [
        {
          "args": [],
          "hook": {
            "code": {
              "file": "githubc.js",
              "kind": "source",
              "target": "someHook",
            },
            "runtimeName": "MyRuntime",
          },
          "kind": "model-hook",
          "name": "description",
          "refKey": "Org.description",
        },
      ],
      "kind": "model",
      "name": "Org",
      "queries": [],
      "refKey": "Org",
      "references": [],
      "relations": [],
    },
  ],
  "populators": [],
  "resolveOrder": [
    "Org",
    "Org.name",
    "Org.description",
  ],
  "runtimes": [
    {
      "default": true,
      "name": "MyRuntime",
      "sourcePath": "some/path/to/file",
      "type": "node",
    },
    {
      "default": false,
      "name": "$GAUDI_INTERNAL",
      "sourcePath": "./internalExecutionRuntime",
      "type": "node",
    },
  ],
}
`;

exports[`compose hooks defaults to the default execution runtime when hook runtime is empty 1`] = `
{
  "actions": [
    {
      "alias": "org",
      "changeset": [
        {
          "name": "name",
          "setter": {
            "args": [],
            "hook": {
              "code": {
                "file": "hooks.js",
                "kind": "source",
                "target": "randomSlug",
              },
              "runtimeName": "MyRuntime",
            },
            "kind": "fieldset-hook",
          },
        },
      ],
      "kind": "create-one",
      "model": "Org",
      "select": [],
      "targetPath": [
        "org",
      ],
    },
  ],
  "authSelect": [],
  "authorize": undefined,
  "fieldset": {
    "kind": "record",
    "nullable": false,
    "record": {},
  },
  "kind": "create",
  "parentContext": [],
  "response": [
    {
      "alias": "id",
      "kind": "field",
      "name": "id",
      "namePath": [
        "Org",
        "id",
      ],
      "refKey": "Org.id",
    },
    {
      "alias": "name",
      "kind": "field",
      "name": "name",
      "namePath": [
        "Org",
        "name",
      ],
      "refKey": "Org.name",
    },
  ],
  "target": {
    "alias": "org",
    "kind": "model",
    "name": "Org",
    "namePath": [
      "Org",
    ],
    "refKey": "Org",
    "retType": "Org",
    "select": [],
  },
}
`;
